hail_scripts/v01/utils/shell_utils.py

```diff
@@ -72,13 +74,28 @@ def run(command,
 
     # pipe output to log
     p = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, env=full_env, bufsize=1)
+    line_buffer = StringIO.StringIO()
     log_buffer = StringIO.StringIO()
-    for line in iter(p.stdout.readline, ''):
-        log_buffer.write(line)
-        if verbose:
-            logger.info(line.strip('\n'))
-    #p.stdout.close()
-
+    previous_is_slash_r = False
+    while True:
+        out = p.stdout.read(1)
+        if out == '' and p.poll() is not None:
+            break
+        if out != '':
+            log_buffer.write(out)
+            if verbose:
+                line_buffer.write(out)
+                if out.endswith('\r') or (out.endswith('\n') and previous_is_slash_r):
```

Comment:
The `universal_newlines` argument to `subprocess.Popen` might simplify this.

https://docs.python.org/2/library/subprocess.html#frequently-used-arguments

Reply:
I tried using universal_newlines but it didn't work correctly for either progress bars like tqdm or  regular line text in some cases.